crates/lsp/src/workspaces.rs

```diff
@@ -0,0 +1,174 @@
+        // to propagate an error here because we don't want to prevent the server from
+        // starting up entirely.
+        // TODO: This is one place it would be nice to show a toast notification back
+        // to the user, but we probably need to add support to the Aux thread for that?
+        for workspace_folder in workspace_folders {
+            if let Err(error) =
+                resolver.open_workspace_folder(&workspace_folder.uri, fallback.clone())
+            {
+                tracing::error!(
+                    "Failed to load workspace settings for '{uri}':\n{error}",
+                    uri = workspace_folder.uri,
+                    error = error
+                );
+            }
+        }
+
+        resolver
+    }
+
+    pub(crate) fn open_workspace_folder(
+        &mut self,
+        url: &Url,
+        fallback: Settings,
+    ) -> anyhow::Result<()> {
+        let path = match Self::url_to_path(url)? {
+            Some(path) => path,
+            None => {
+                tracing::warn!("Ignoring non-file workspace URL: {url}");
+                return Ok(());
+            }
+        };
+
+        let mut settings_resolver = SettingsResolver::new(fallback);
+        settings_resolver.load_from_paths(&[&path])?;
+
+        tracing::trace!("Adding workspace settings: {}", path.display());
+        self.path_to_settings_resolver.add(&path, settings_resolver);
+
+        Ok(())
+    }
+
+    pub(crate) fn close_workspace_folder(
+        &mut self,
+        url: &Url,
+    ) -> anyhow::Result<Option<SettingsResolver>> {
+        match Self::url_to_path(url)? {
+            Some(path) => {
+                tracing::trace!("Removing workspace settings: {}", path.display());
+                Ok(self.path_to_settings_resolver.remove(&path))
+            }
+            None => {
+                tracing::warn!("Ignoring non-file workspace URL: {url}");
+                Ok(None)
+            }
+        }
+    }
+
+    /// Return the appropriate [`Settings`] for a given document [`Url`].
+    pub(crate) fn settings_for_url(&self, url: &Url) -> &Settings {
+        if let Ok(Some(path)) = Self::url_to_path(url) {
+            return self.settings_for_path(&path);
+        }
+
+        // For `untitled` schemes, we have special behavior.
+        // If there is exactly 1 workspace, we resolve using a path of
+        // `{workspace_path}/untitled` to provide relevant settings for this workspace.
+        if url.scheme() == "untitled" && self.path_to_settings_resolver.len() == 1 {
+            tracing::trace!("Using workspace settings for 'untitled' URL: {url}");
+            let workspace_path = self.path_to_settings_resolver.keys().next().unwrap();
+            let path = workspace_path.join("untitled");
+            return self.settings_for_path(&path);
+        }
+
+        tracing::trace!("Using default settings for non-file URL: {url}");
+        self.path_to_settings_resolver.fallback().fallback()
+    }
```

Comment:
Tricky little behavior here with untitled files

If there aren't any workspaces open, then the untitled file can't be associated with any `air.toml`

If there are >1 workspaces open, we can't make a reasonable guess about which workspace's `air.toml` to try and associate the untitled file with

If there is exactly 1 workspace open (the 95% case), then we can make a reasonable guess that we should use that workspace's `air.toml` with untitled files